### ISSUE TITLE
[Proposal] Support Range and Tuples on Attributes

### ISSUE THREAD

**Issue by eyalsk** [dotnet/csharplang#3535]
# Support Range and Tuples on Attributes

* [x] Proposed
* [ ] Prototype: Not Started
* [ ] Implementation: Not Started
* [ ] Specification: Not Started

## Summary
[summary]: #summary

Add support for `Range` and `ValueTuple` and their respective syntax on attributes.

## Motivation
[motivation]: #motivation

Due to lack of support for tuples and ranges, in my unit tests I'm using arrays and strings like so:
```
[TestCase("[A]\n[2]", "Range: 0..3; Value: A", "Range: 4..7; Value: 2")]
```
There are couple of issues with it:
1. **Performance**: You have to either parse the string or use string format to compare the values and that's pretty slow.
2. **Readability**: It's not immediate what part of the string is the data that is being tested so you have to inspect the test and it's pretty error-prone too.
3. **Expressiveness**: Range and tuples are already supported by the language and the framework and are pretty common so there's no reason I should go out of my way and express them using strings when there are better types to represents the data.

## Detailed design
[design]: #detailed-design

One possible way to implement both of them is to represent the information using arrays which are already supported by the language.

So say we have the following tuple `(start, end)` we can represent this like so:
```
new object[] { typeof(ValueTuple<int, int>), start, end }
```

And the same idea goes for range say we have `start..end` we can represent this like so:
```
new object[] { typeof(Range), start, end }
```

In terms of syntax and semantics there shouldn't be any differences, this is all just an implementation detail.

We've discussed this approach over Gitter and some people had concerns and stuff so it's better to have a proper discussion about it and maybe there's a much better way to do it by adding native support for this depends on goals and whatnot, maybe it might be even easier to implement it, dunno.

## Drawbacks
[drawbacks]: #drawbacks

Don't think there are any.

## Alternatives
[alternatives]: #alternatives

As stated above, we can use strings and/or arrays to represent the data but it can be cumbersome and error-prone.

**Comment by HaloFour** [comment]
I'd rather see the CLR modified to properly support more types (or any blittable type) vs. having some kind of hack using arrays.

**Comment by eyalsk** [comment]
@HaloFour I agree, probably would be a lot better but I don't know how much work each approach requires so I thought to provide an alternative and create an issue for it.

**Comment by CyrusNajmabadi** [comment]
I agree with @HaloFour. Attributes are about a metadata exchange format to pass along information. If it's not an internal compiler detail, then it should be a first class concept. This would be the latter, so I'd want the clr to support this.

**Comment by eyalsk** [comment]
@CyrusNajmabadi Should I remove the suggested implementation part then and just have the issue focus on 1st class support?

**Comment by CyrusNajmabadi** [comment]
I think the impl part is fine.  i'm just saying that i don't personally agree with the idea that we should be comign up with a way to encode first-class concepts like this into attributes.  others may disagree :)

**Comment by jnm2** [comment]
As discussed [here](https://github.com/dotnet/csharplang/issues/3439#issuecomment-626233136), I'd really like to be able to do things like `MyStruct.SomeStaticMethod(new DateTime(2020, 1, 1))` .

Attributes let you specify enums, but when DUs come along and I want to parameterize enums, I won't be able to use them as attribute values.

**Comment by jnm2** [comment]
@CyrusNajmabadi 

> i'm just saying that i don't personally agree with the idea that we should be comign up with a way to encode first-class concepts like this into attributes.

There isn't a great option for providing test data to test methods today. It's bulky and unpleasant in all the testing frameworks. Letting attributes encode first-class concepts would be one way to improve this, but I'd like to see an alternative suggestion that addresses this underlying need before dismissing the attribute route.
